### PR TITLE
Fixed #2223: Implement Predictive Index and add more unit tests

### LINKED ISSUE #2223: Predictive Query API
**PRD**
1. https://docs.google.com/document/d/1ydyIaXLYv9871qWDQYJontaee4xTZgOxMJuqXtzIv1Q/edit

**API Spec**
1. https://github.com/couchbaselabs/couchbase-lite-apiv2/pull/229

**LiteCore API**
1. https://github.com/couchbase/couchbase-lite-core/blob/master/C/include/c4PredictiveQuery.h
2. https://github.com/couchbase/couchbase-lite-core/blob/master/C/include/c4Query.h#L194
3. https://github.com/couchbase/couchbase-lite-core/blob/3927e66f5685ef1d7b320f0a55799a0236610ec8/LiteCore/tests/PredictiveQueryTest.cc

### PR DESCRIPTION
**Pull request by pasin**
* Implemented Predictive Index API
* Updated and added more unit tests

#2223

### CODE REVIEW

File: Objective-C/Internal/CBLValueExpression.m [hunk 1]
```diff
@@ -82,6 +82,7 @@ static id dictionaryAsJSON(NSDictionary* dict) {
 
 static id arrayAsJSON(NSArray* array) {
     NSMutableArray *json = [NSMutableArray arrayWithCapacity: array.count];
```
Comment (jayahariv) [inline Objective-C/Internal/CBLValueExpression.m]: better to allocate the extra array operation space too. 

```suggestion
    NSMutableArray *json = [NSMutableArray arrayWithCapacity: array.count + 1];
```

File: Objective-C/CBLPredictiveIndex.m [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+#import "CBLQueryExpression+Internal.h"
+#import "CBLPredictiveIndex+Internal.h"
+
+@implementation CBLPredictiveIndex {
+    NSString* _model;
+    CBLQueryExpression* _input;
+    NSArray<NSString*>* _properties;
+}
+
+- (instancetype) initWithModel: (NSString*)model
```
Comment (jayahariv) [inline Objective-C/CBLPredictiveIndex.m]: an extra line? 
```suggestion

- (instancetype) initWithModel: (NSString*)model
```

File: Objective-C/Internal/CBLValueExpression.m (hunk 1 above)
Comment (pasin) [reply Objective-C/Internal/CBLValueExpression.m]: Thanks for catching that.

File: Objective-C/CBLPredictiveIndex.m (hunk 2 above)
Comment (pasin) [reply Objective-C/CBLPredictiveIndex.m]: Sure.

File: Objective-C/Tests/PredictiveQueryTest.m [hunk 3]
```diff
@@ -95,100 +98,146 @@ - (void) testRegisterAndUnregisterModel {
...
+    }];
+    AssertEqual(numRows, 1u);
+    
+    [CBLDatabase.prediction unregisterModelWithName: model];
 }
 
-- (void) testQueryValueFromDictionaryResult {
-    [self createDocumentWithNumbers: @[@1, @2, @3, @4, @5]];
-    [self createDocumentWithNumbers: @[@6, @7, @8, @9, @10]];
+- (void) testPredictionInpuOutput {
```
Comment (jayahariv) [inline Objective-C/Tests/PredictiveQueryTest.m]:
```suggestion
- (void) testPredictionInputOutput {
```

**Comment by pasin** [comment]
I fixed the PR per review comment. Please take a look when you have time.

**Pull request merged**